File: models/news.py
from typing import Dict, Union

class NewsData:
    """Data model for news articles"""
    def __init__(self, raw_data: Union[Dict, str]):
        if isinstance(raw_data, dict):
            self.headline = raw_data.get("headline", raw_data.get("title", "No headline"))
            self.description = raw_data.get("description", "")
            self.web_url = raw_data.get("web_url") or raw_data.get("links", {}).get("web", {}).get("href", "")
            self.mobile_url = raw_data.get("mobile_url") or raw_data.get("links", {}).get("mobile", {}).get("href", "")
            self.published = raw_data.get("published", "")
            self.byline = raw_data.get("byline", "")
        else:
            self.headline = str(raw_data)
            self.description = ""
            self.web_url = ""
            self.mobile_url = ""
            self.published = ""
            self.byline = ""

    def get_display_text(self) -> str:
        display_text = self.headline
        if self.byline:
            display_text += f" | {self.byline}"
        if self.published:
            display_text += f" | {self.published[:10]}"
        if self.description:
            display_text += f"\n  {self.description[:100]}{'...' if len(self.description) > 100 else ''}"
        return display_text

    def has_web_url(self) -> bool:
        return bool(self.web_url)

File: SportsScoresApp/SportsScores/Models/News.swift
//
//  News.swift
//  SportsScores
//
//  Data model for ESPN news articles.
//

import Foundation

struct NewsItem: Identifiable {
    let id: String
    let headline: String
    let description: String?
    let byline: String?
    let published: Date?
    let articleURL: URL?
    let imageURL: URL?

    init(from api: NewsAPIResponse.APIArticle) {
        // ESPN returns id as an integer; convert to String for Identifiable.
        self.id          = api.id.map { String($0) } ?? UUID().uuidString
        self.headline    = api.headline
        self.description = api.description
        self.byline      = api.byline
        self.published   = api.published
        self.articleURL  = api.links?.web?.href.flatMap { URL(string: $0) }
        self.imageURL    = api.images?.first?.url.flatMap { URL(string: $0) }
    }
}

// MARK: - API Response Models

struct NewsAPIResponse: Codable {
    let articles: [APIArticle]

    struct APIArticle: Codable {
        /// ESPN returns article IDs as integers (e.g. 48040023).
        let id: Int?
        let headline: String
        let description: String?
        let byline: String?
        let published: Date?
        let links: Links?
        let images: [NewsImage]?

        struct Links: Codable {
            let web: WebLink?

            struct WebLink: Codable {
                let href: String?
            }
        }

        struct NewsImage: Codable {
            let url: String?
        }
    }
}
